### PR TITLE
Add assessment access flag for hiding grades

### LINKED ISSUE #2181: Add "don't display the student score under any circumstances" feature for assessments
Requested by @cheeren and @lkwagner.

This should replicate the experience of a pen-and-paper exam, where you answer all the questions (saving answers as many times as you want) but it's graded later by the professor.

### PR DESCRIPTION
**Pull request by falkirks**
This resolves #2181. 

Added a flag (showClosedAssessmentGrade) that can be set to false in conjunction with showClosedAssessment to hide all grade information from a student. Tested with a new JS test file.

### CODE REVIEW

File: api/v1/endpoints/queries.sql [hunk 1]
```diff
@@ -99,6 +99,7 @@ WITH object_data AS (
         aar.role,
         aar.seb_config,
         aar.show_closed_assessment,
+        aar.show_closed_assessment_grade,
```
Comment (mwest1066) [inline api/v1/endpoints/queries.sql:102]: I'm sorry to ask this, but could we rename this to `show_closed_assessment_score`? Within PrairieLearn we have generally been using the term "score" as the noun for the numerical score on each question/assessment, and we use "grade" for the verb of the action that computes scores. I'm not going to attempt to justify this choice of terminology 😄

**Review by mwest1066 — COMMENTED** [review]
Fantastic PR! I'm very impressed at the way you implemented this in exactly the same code style and methodology as existing functionality. It made reviewing the PR fast and pleasant!

Two other items:
* Documentation. Could we add `showClosedAssessmentScore` to the table on the [allowAccess](https://github.com/PrairieLearn/PrairieLearn/blob/master/docs/accessControl.md) page and also add a section like [the one for `showClosedAssessment`](https://github.com/PrairieLearn/PrairieLearn/blob/master/docs/accessControl.md#showinghiding-closed-assessments)?
* Students can still see their total scores for all assessments on the student [Gradebook](https://github.com/PrairieLearn/PrairieLearn/tree/master/pages/studentGradebook) page. To hide the score there as well, that page is going to need to call `authz_assessment()` like is done on `studentAssessments`: https://github.com/PrairieLearn/PrairieLearn/blob/9606b8d25c9bf284899c8e9598882bddbab2c1a1/pages/studentAssessments/studentAssessments.sql#L30 For the Gradebook page we will need to show all assessments whether they are `authorized` or not, but hide the score for those that have `show_closed_assessment_score = FALSE`.

File: api/v1/endpoints/queries.sql (hunk 1 above)
Comment (falkirks) [reply api/v1/endpoints/queries.sql]: Done!

File: pages/studentAssessments/studentAssessments.ejs [hunk 2]
```diff
@@ -74,8 +74,12 @@
                 <a href="<%= urlPrefix %><%= row.link %>" class="btn btn-primary btn-sm" role="button">New instance</a>
                 <% } else { // multiple_instance_header %>
                 <% if (row.assessment_instance_id) { %>
+                <% if(row.show_closed_assessment_grade) { %>
                 <%- include('../partials/scorebar', {score: row.assessment_instance_score_perc}) %>
                 <% } else { %>
+                Score withheld.
```
[the remark this answers is not in the document]
Comment (falkirks) [reply pages/studentAssessments/studentAssessments.ejs]: Done!

**Comment by falkirks** [comment]
@mwest1066 I'm not sure about the gradebook page. I think I have already hidden the scores in the gradebook page in this PR. Is there something I am missing?

Otherwise I have integrated the rest of your feedback

**Comment by mwest1066** [comment]
Thanks for all the updates! Looks great.

> @mwest1066 I'm not sure about the gradebook page. I think I have already hidden the scores in the gradebook page in this PR. Is there something I am missing?

I was referring to the page shown below. This is [`pages/studentGradebook`](https://github.com/PrairieLearn/PrairieLearn/tree/master/pages/studentGradebook). This shows all student scores unconditionally, independent of student access controls.

To also hide the score on this page I think we'll need to change the query to also call `authz_assessment` like we do for the main student assessments page: https://github.com/PrairieLearn/PrairieLearn/blob/6084b9e6dd1756bfb5cc62dfb4bf2b1f18f49efc/pages/studentAssessments/studentAssessments.sql#L99

The difference here would be that we should show the score on the gradebook page if either: (1) access is denied, or (2) access is allowed and `showClosedAssessmentScore IS NOT FALSE`. I think that will preserve the current behavior (all assessment instances) for courses not using `showClosedAssessmentScore`, and will only hide those assessment instances with `showClosedAssessmentScore` specifically set for courses using it. What do you think? Would that be reasonable?

image

File: pages/studentGradebook/studentGradebook.ejs [hunk 3]
```diff
@@ -50,8 +50,12 @@
               </td>
               <td class="text-center align-middle">
                 <% if (!row.multiple_instance_header) { %>
+                <% if(!row.show_closed_assessment_score) { %>
+                Score not shown
```
Comment (falkirks) [inline pages/studentGradebook/studentGradebook.ejs:54]: @mwest1066 This is where I hide the grades in the gradebook.

**Comment by falkirks** [comment]
@mwest1066 I really think I am doing this already. Perhaps there is a bug and you aren't seeing it, but it definitely hides on my end. I will attach a picture later today.

**Comment by mwest1066** [comment]
> @mwest1066 I really think I am doing this already. Perhaps there is a bug and you aren't seeing it, but it definitely hides on my end. I will attach a picture later today.

Many apologies! Thanks for being so patient with me and pointing out exactly what was going on. I'm not sure how I overlooked your code that did exactly what I was requesting 😄

**Pull request merged**